Added models for the missing gets and fgets IO functions

lucasccordeiro opened the pull request:
This PR should add models for the missing `gets` and `fgets` IO functions from the `stdio.h` library.

lucasccordeiro committed:
[c2goto] Added models for missing io functions

Added operational models for the missing gets and fgets
IO functions from the stdio library.

lucasccordeiro committed:
[regression] Added regression for IO functions

Added three regression tests for the gets and fgets IO
functions from the stdio.h library. In particular, these
regression tests have been added to the unix folder.

fbrausse commented on src/c2goto/library/io.c:
```diff
@@ -61,8 +71,10 @@ __ESBMC_HIDE:;
-  _Bool error;
-  return error ? 0 : str;
+  int i = 0;
+  for(i = 0; i < size; i++)
```
`fgets` is supposed to read <= size-1 bytes and to append a terminating `\0`.

lucasccordeiro replied:
Fixed.

fbrausse commented on src/c2goto/library/io.c:
```diff
@@ -61,8 +71,19 @@ __ESBMC_HIDE:;
+    return NULL;
+
+  // add nondeterministic characters to str
+  for(int i = 0; i <= (size - 1); i++)
```
Should also stop after a new-line character is encountered (and on EOF).

lucasccordeiro replied:
Fixed.

lucasccordeiro committed:
[regression] Updated regression tests

Updated regression tests to consider the newly introduced
operational model for gets and fgets.

fbrausse commented on src/c2goto/library/io.c:
```diff
@@ -58,11 +68,43 @@ __ESBMC_HIDE:;
+      str[i] = '\0';
+      break;
+    }
+    str[i] = nondet_char();
```
This should store `character`.

lucasccordeiro replied:
Fixed.

fbrausse replied:
`nondet_char()` could be `'\n'` here, which by contract of `fgets` is not allowed since it would have stopped after.

lucasccordeiro replied:
Fixed.

fbrausse commented on src/c2goto/library/io.c:
```diff
@@ -58,11 +68,43 @@ __ESBMC_HIDE:;
+// into the string pointed to by str. It stops when either (n-1)
+// characters are read, the newline character is read, or the
+// end-of-file is reached, whichever comes first.
+// source: https://www.cplusplus.com/reference/cstdio/fgets/
```
It goes saying
> A newline character makes fgets stop reading, but it is considered a valid character by the function and included in the string copied to str.

lucasccordeiro replied:
Improved the documentation as suggested.

fbrausse commented on src/c2goto/library/io.c:
```diff
@@ -58,11 +68,43 @@ __ESBMC_HIDE:;
+    // produce non-deterministic character
+    char character = nondet_char();
+    // stop if we get a new line character or an EOF
+    if(character == '\n' || character == EOF)
```
As long as our `feof()` just returns `nondet_int()` and does not record the EOF/ferror bits in the `FILE` itself, it may not make a lot of sense to test this condition here. But we could end the loop early based on non-determinism. Note that `EOF` is `(int)-1`

lucasccordeiro replied:
Fixed.

fbrausse reviewed (changes requested):
Should the regression tests be moved out of `unix2`? They cover C stdlib functionality.

lucasccordeiro commented:
> Should the regression tests be moved out of unix2? They cover C stdlib functionality.

Done.

mikhailramalho commented on src/c2goto/library/io.c:
```diff
@@ -37,6 +37,16 @@ __ESBMC_HIDE:;
+  size_t size = nondet_uint();
+  for(i = 0; i < size; i++)
+    buf[i] = nondet_char();
+  return buf;
```
The last element should also be a null character: https://www.cplusplus.com/reference/cstdio/gets/

lucasccordeiro replied:
Fixed.

mikhailramalho commented on src/c2goto/library/io.c:
```diff
@@ -58,11 +68,53 @@ __ESBMC_HIDE:;
+    {
+      // we end the loop based on non-determinism
+      str[i] = '\0';
+      break;
```
are we adding '\0' twice?

lucasccordeiro replied:
Fixed.

mikhailramalho commented on src/c2goto/library/io.c:
```diff
@@ -58,11 +68,53 @@ __ESBMC_HIDE:;
+  }
+
+  // append a terminating \0
+  str[size - 1] = '\0';
```
I think this should be `i` instead of `size` so we don't add '\0' twice.

lucasccordeiro replied:
Fixed.

mikhailramalho reviewed (commented):
Small comments to check, otherwise LGTM.

fbrausse commented on src/c2goto/library/io.c:
```diff
@@ -37,6 +37,16 @@ __ESBMC_HIDE:;
+  size_t i = 0;
+  size_t size = nondet_uint();
+  for(i = 0; i < size; i++)
+    buf[i] = nondet_char();
```
From the documentation: "Reads [...] and stores [...] until a newline character or the end-of-file is reached." We should ensure these invariants.

lucasccordeiro replied:
Fixed.

fbrausse commented on src/c2goto/library/io.c:
```diff
@@ -37,6 +37,30 @@ __ESBMC_HIDE:;
+  for(i = 0; i < size; i++)
+  {
+    char character = nondet_char();
+    if(character == '\n' || character == EOF)
```
Remove `|| character == EOF`: this would fail when `(unsigned char)character == 0xff`. As `size` is already non-det. here, it should be fine not to check for EOF (until we properly model that).

lucasccordeiro replied:
Fixed.

mikhailramalho commented:
I wonder if we should spend time implementing support for `forall` for the solvers that support it, as it would simplify these models considerably.

lucasccordeiro commented:
> I wonder if we should spend time implementing support for forall for the solvers that support it, as it would simplify these models considerably.

@mikhailramalho: do you remember which solvers support `forall`? Z3, CVC4 and Yices?

fbrausse reviewed (approved):
So regression tests remain in unix2? Otherwise LGTM

mikhailramalho commented:
Yeah, I think so.

We used to have a z3 implementation but it got removed a long time ago.

Em qua., 22 de dez. de 2021 às 09:39, Lucas Cordeiro <
***@***.***> escreveu:

> I wonder if we should spend time implementing support for forall for the
> solvers that support it, as it would simplify these models considerably.
>
> @mikhailramalho <https://github.com/mikhailramalho>: do you remember
> which solvers support forall? Z3, CVC4 and Yices?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/573#issuecomment-999584201>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJHZZCAAM4JHB4FRIVX3USHIJJANCNFSM5KQNUMWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.

lucasccordeiro commented:
> So regression tests remain in unix2? 

I'm unsure how is our support of the C stdlib on Windows.

@rafaelsamenezes is currently on vacation, who is our guru on Windows; he'll be back to work only on January 4th, 2022.

shmarovfedor reviewed (approved):
LGTM
